lib/plug/csrf_protection.ex

```diff
@@ -103,17 +103,34 @@ defmodule Plug.CSRFProtection do
 
   def init(opts), do: opts
 
-  def call(conn, _opts) do
+  def call(conn, opts) do
+    opts = verify_opts opts
     csrf_token = get_session(conn, "_csrf_token")
     Process.put(:plug_csrf_token, csrf_token)
-
     if not verified_request?(conn, csrf_token) do
-      raise InvalidCSRFTokenError
+      if opts[:with] == :exception do
```

Comment:
The advantage of using `opts[:clear_session]` is that you don't have to provide a default value for it.There's just the default error raising or clearing of the session. Error-raising is the default behavior and if you want to clear session, just set `:clear_session` to true. 

Also, if you just check if `opts[:clear_session]` has a truthy value, you can get rid of the validation of the value passed to `opts[:with]`.

There'll be less code and the implementation is simpler.